activiti-core/activiti-engine/src/main/java/org/activiti/engine/impl/bpmn/behavior/MultiInstanceActivityBehavior.java

```diff
@@ -449,11 +452,21 @@ protected void updateResultCollection(DelegateExecution childExecution,
             } else {
                 resultCollection = new ArrayList<>();
             }
-            resultCollection.add(childExecution.getVariable(getOutputDataItem()));
+            resultCollection.add(getResultElementItem(childExecution));
             setLoopVariable(miRootExecution, getLoopDataOutputRef(), resultCollection);
         }
     }
 
+    private Object getResultElementItem(DelegateExecution childExecution) {
+        CommandContext commandContext = Context.getCommandContext();
+        if (commandContext.getCommand() instanceof CompleteTaskCmd) {
+            Map<String, Object> taskVariables = ((CompleteTaskCmd) commandContext
```

Comment:
Why only for `CompleteTaskCmd` ?

Reply:
Good point! Actually, user tasks follow a different path because the variables are attached directly to the task entity (which is deleted during the execution of CompleteTaskCmd). For the other cases, the variables will be available in the execution. I preferred to not change the current mechanism that we have for variables on user tasks as it can cause some unwanted behavior.

Reply:
I added a comment explaining this.